File: java/com/purescraps/purehtml/tests/Specs.java
package com.purescraps.purehtml.tests;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.purescraps.purehtml.PureHTML;
import com.purescraps.purehtml.configs.Config;
import com.purescraps.purehtml.configs.ConfigFactory;
import org.everit.json.schema.Schema;
import org.everit.json.schema.loader.SchemaLoader;
import org.json.JSONObject;
import org.json.JSONStringer;
import org.yaml.snakeyaml.Yaml;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.*;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class Specs {

    public static void main(String[] args) {
        // Directories
        Path startPath = Paths.get("specs");
        try (Stream<Path> filePathStream = Files.walk(startPath)) {
            filePathStream.filter(Files::isRegularFile) // Filter out directories, leaving only files
                    .forEach(path -> {
                        // Read the file content
                        try {
                            //If validation Success, continue
                            if (validate(path.toString())) {
                                String content = Files.readString(path);
                                ObjectMapper yamlMapper = new ObjectMapper(new YAMLFactory());
                                Map<String, Object> yamlMap = yamlMapper.readValue(content, new TypeReference<>() {
                                });
                                @SuppressWarnings("unchecked")
                                List<Map<String, Object>> specs = (List<Map<String, Object>>) yamlMap.get("specs");

                                for (Map<String, Object> spec : specs) {
                                    String html = spec.get("input") != null ? spec.get("input").toString() : "";
                                    @SuppressWarnings("unchecked")
                                    Map<String, Object> configuration = (Map<String, Object>) spec.get("configuration");
                                    Config config = ConfigFactory.fromYAML(configuration);
                                    Object answer = PureHTML.extract(config, html, configuration);
                                    @SuppressWarnings("unchecked")
                                    List<String> expected = (List<String>) spec.get("expected");
                                    ObjectMapper objectMapper = new ObjectMapper();
                                    Object outputObject = objectMapper.readValue(answer.toString(), Object.class);
                                    Object expectedObject = objectMapper.readValue(JSONStringer.valueToString(expected), Object.class);
                                    if (!outputObject.equals(expectedObject)) {
                                        System.out.println("File: " + path);
                                        System.out.println("Description: " + spec.get("description"));
                                        System.out.println("Extracted Values: " + outputObject);
                                        System.out.println("Expected Values : " + expectedObject);
                                    } else {

                                        //comment below to not see successful ones.
                                        System.out.println("Success: " + path );
                                    }
                                }
                            }
                        } catch (IOException e) {
                            System.err.println("Error reading file: " + path + " - " + e.getMessage());
                        }
                    });
        } catch (IOException e) {
            System.err.println("Error walking file tree: " + e.getMessage());
        }
    }

    private static boolean validate(String yamlFilePath){
        //Below is the schema path from the repository folder, which is just file name itself.
        String schemaPath = "spec.schema.yaml";
        // Load the YAML schema and yamlFile into a Java object and validate
        try{
            Yaml yaml = new Yaml();
            FileInputStream yamlInputStream = new FileInputStream(schemaPath);
            Object schemaObject = yaml.load(yamlInputStream);

            ObjectMapper objectMapper = new ObjectMapper();
            String jsonSchema = objectMapper.writeValueAsString(schemaObject);

            yaml = new Yaml();
            FileInputStream yamlInputStream2 = new FileInputStream(yamlFilePath);
            Object yamlObject = yaml.load(yamlInputStream2);

            objectMapper = new ObjectMapper();
            String jsonContent = objectMapper.writeValueAsString(yamlObject);

            JSONObject jsonSchemaObject = new JSONObject(jsonSchema);
            Schema schema = SchemaLoader.load(jsonSchemaObject);

            JSONObject jsonObject = new JSONObject(jsonContent);
            schema.validate(jsonObject);  // This will throw an exception if the validation fails
            return true;
        }
        catch(FileNotFoundException e){
            System.out.println(e.getMessage());
            return false;

        } catch (JsonProcessingException e) {
            System.out.println("File: " +yamlFilePath + ": " + e.getMessage() );
            return false;
        }

    }

}

File: python/purehtml/tests/Specs.py
import json
import yaml
from pathlib import Path
from jsonschema import validate, ValidationError

from purehtml.PureHTML import PureHTML
from purehtml.configs.ConfigFactory import ConfigFactory


def run():

    # Directory containing the YAML specs
    specs_dir = Path("../specs")
    try:
        for file_path in specs_dir.rglob("*"):

            if file_path.is_file():  # Process only files
                try:
                    # If validation is successful, continue processing
                    if validate_file(file_path):
                        with file_path.open("r", encoding="utf-8") as file:
                            content = file.read()

                        yaml_map = yaml.safe_load(content)
                        specs = yaml_map.get("specs", [])

                        for spec in specs:
                            html = spec.get("input", "")
                            configuration = spec.get("configuration", {})
                            # Create the config object
                            config = ConfigFactory.from_dict(configuration)

                            # Extract the answer using PureHTML
                            answer = PureHTML.extract(config, html, configuration)

                            # Get the expected values
                            expected = spec.get("expected", [])
                            expected_object = str(json.loads(json.dumps(expected)))
                            if answer == "[]":
                                answer = "None"
                            if str(answer) != str(expected_object):
                                print(f"Incorrect values at: {file_path}")
                                print(f"Description: {spec.get('description')}")
                                print(f"Extracted Values: {answer}")
                                print(f"Expected Values : {expected_object}")
                            else:
                                # Uncomment the line below to skip successful outputs
                                print(f"Success: {file_path}")
                                #print(f"Extracted Values: {answer}")
                                #print(f"Expected Values : {expected_object}")

                except Exception as e:
                    print(f"Error reading file: {file_path} - {str(e)}")

    except Exception as e:
        print(f"Error walking file tree: {str(e)}")


def validate_file(yaml_file_path: Path) -> bool:
    """
    Validate a YAML file against the schema.
    :param yaml_file_path: Path to the YAML file.
    :return: True if valid, False otherwise.
    """
    schema_path = Path("../spec.schema.yaml")

    try:
        # Load the schema
        with schema_path.open("r", encoding="utf-8") as schema_file:
            schema_object = yaml.safe_load(schema_file)

        # Convert schema to JSON
        schema_json = json.loads(json.dumps(schema_object))

        # Load the YAML content
        with yaml_file_path.open("r", encoding="utf-8") as yaml_file:
            yaml_content = yaml.safe_load(yaml_file)

        # Validate the YAML content against the schema
        validate(instance=yaml_content, schema=schema_json)
        return True

    except FileNotFoundError as e:
        print(e)
        return False

    except ValidationError as e:
        print(f"Validation Error: {e.message} - File: {yaml_file_path}")
        return False

    except Exception as e:
        print(f"File: {yaml_file_path} - {str(e)}")
        return False


